crates/pipeline_manager/src/auth.rs

```diff
@@ -0,0 +1,368 @@
+use std::{collections::HashMap, env};
+
+use actix_web::dev::ServiceRequest;
+use actix_web_httpauth::extractors::{
+    bearer::{BearerAuth, Config},
+    AuthenticationError,
+};
+use cached::SizedCache;
+use jsonwebtoken::{decode, decode_header, Algorithm, DecodingKey, TokenData, Validation};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+
+// Authorization using a bearer token. This is strictly used for authorizing
+// users, not machines.
+pub(crate) async fn auth_validator(
+    configuration: AuthConfiguration,
+    req: ServiceRequest,
+    credentials: BearerAuth,
+) -> Result<ServiceRequest, (actix_web::error::Error, ServiceRequest)> {
+    let token = credentials.token();
+    let token = match configuration.provider {
+        Provider::AwsCognito(_) => decode_aws_cognito_token(token, configuration).await,
+    };
+    match token {
+        Ok(_) => Ok(req),
+        Err(e) => {
+            let config = req.app_data::<Config>().cloned().unwrap_or_default();
+            Err((
+                AuthenticationError::from(config)
+                    .with_error_description(e.to_string())
+                    .into(),
+                req,
+            ))
+        }
+    }
+}
+
+#[derive(Debug)]
+enum Claim {
+    AwsCognito(TokenData<AwsCognitoClaim>),
+}
+
+#[derive(Clone)]
+pub(crate) enum Provider {
+    AwsCognito(String), // The argument is the URL to use for fetching JWKs
+}
+
+pub(crate) fn aws_auth_config() -> AuthConfiguration {
+    let mut validation = Validation::new(Algorithm::RS256);
+    let audience =
+        env::var("OAUTH_CLIENT_ID").expect("Missing environment variable OAUTH_CLIENT_ID");
+    let iss = env::var("OAUTH_ISSUER").expect("Missing environment variable OAUTH_ISSUER");
+    let jwk_uri = env::var("OAUTH_JWK_URI").expect("Missing environment variable OAUTH_JWK_URI");
+    validation.set_audience(&[audience]);
+    validation.set_issuer(&[iss]);
+    AuthConfiguration {
+        provider: Provider::AwsCognito(jwk_uri),
+        validation,
+    }
+}
+
+#[derive(Clone)]
+// Expected issuer and client_id for each authentication request
+pub(crate) struct AuthConfiguration {
+    pub provider: Provider,
+    pub validation: Validation,
+}
+
+// https://docs.aws.amazon.com/cognito/latest/developerguide/amazon-cognito-user-pools-using-the-access-token.html
+#[derive(Clone, Debug, Serialize, Deserialize)]
+struct AwsCognitoClaim {
```

Comment:
Who provides AwsCognitoClaim? the user or aws? maybe some high-level comment that explains would be good

Reply:
It is given to the client by AWS.